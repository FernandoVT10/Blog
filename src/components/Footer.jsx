import Link from "next/link";
import Input from "./form/Input";
import { useState } from "react";

import "../styles/components/footer.scss";

export default () => {
    const [email, setEmail] = useState({value: "", valid: false});

    const suscribe = e => {
        e.preventDefault();

        if(email.valid) {
            alert("holu");
        }
    };

    return (
        <footer className="main-footer container-fluid mt-4">
            <div className="row">
                <div className="col-12 col-sm-6 col-xl-3">
                    <Link href="/">
                        <a className="main-footer__brand">
                            <img
                            src="favicon.ico"
                            width="50"
                            height="50"
                            className="d-inline-block align-top rounded-circle"
                            alt="Icon" />
                            <h3 className="main-footer__title">
                                Fernando Vaca Tamayo
                            </h3>
                        </a>
                    </Link>

                    <div className="mt-4">
                        <h3 className="main-footer__subtitle">
                            Contact Me
                        </h3>

                        <p>
                            <i
                            className="fas fa-envelope main-footer__icon"
                            aria-hidden="true"></i>

                            <span className="main-footer__text">
                                mailexample@gmail.com
                            </span>
                        </p>
                    </div>
                </div>

                <div className="col-12 col-sm-6 col-xl-3">
                    <h3 className="main-footer__subtitle">
                        Social Networks
                    </h3>

                    <a href="https://twitter.com/FernandoVT10" target="_blank">
                        <p>
                            <i
                            className="fab fa-twitter main-footer__icon"
                            aria-hidden="true"></i>

                            <span className="main-footer__text">
                                @FernandoVT10
                            </span>
                        </p>
                    </a>

                    <a href="https://github.com/FernandoVT10" target="_blank">
                        <p>
                            <i
                            className="fab fa-github main-footer__icon"
                            aria-hidden="true"></i>

                            <span className="main-footer__text">
                                FernandoVT10
                            </span>
                        </p>
                    </a>
                </div>

                <div className="col-12 col-sm-6 col-xl-3">
                    <h3 className="main-footer__subtitle">
                        Description
                    </h3>

                    <p className="main-footer__description">
                        This blog is my personal blog xD
                    </p>
                </div>

                <div className="col-12 col-sm-6 col-xl-3">
                    <h3 className="main-footer__subtitle">
                        Suscribe to my blog
                    </h3>

                    <form onSubmit={suscribe}>
                        <Input
                        type="email"
                        placeholder="Enter your email"
                        onChange={setEmail} />

                        <button className="submit-button">
                            Suscribe
                        </button>
                    </form>
                </div>
            </div>
        </footer>
    );
};