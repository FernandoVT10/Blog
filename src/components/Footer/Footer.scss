.main-footer {
    background: var(--primary);
    padding: 30px 10px;

    &__brand {
        display: flex;
        align-items: center;

        &:hover {
            text-decoration: none;
        }
    }

    &__title {
        font-family: 'Pacifico', cursive;
        font-size: 1.25em;
        margin: 0;
        margin-left: 10px;
        color: var(--white);
    }

    &__subtitle {
        color: var(--secondary);
        font-size: 1.25em;
        font-weight: 600;
        margin-top: 1.5em;
    }

    &__icon {
        color: var(--secondary);
    }

    &__text {
        color: var(--white);
        margin-left: 10px;
        font-weight: bold;
    }

    &__description {
        color: var(--white);
        font-weight: bold;
        line-height: 1.8em;
    }

    &__subscription-form-loader {
        margin-top: 40px;
        text-align: center;
        color: var(--secondary);
    }

    @media screen and (min-width: 576px) {
        padding: 30px;

        &__subtitle {
            margin-top: 0;
        }
    }
}