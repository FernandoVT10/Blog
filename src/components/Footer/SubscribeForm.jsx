import ApiController from "../../services/ApiController";

import { useEmailValidation } from "../../hooks/useEmailValidation";

import { useState } from "react";

export default () => {
    const [email, setEmail] = useEmailValidation();

    const [error, setError] = useState("");
    const [success, setSuccess] = useState("");
    const [loading, setLoading] = useState(false);

    const suscribe = e => {
        e.preventDefault();

        if(!email.error) {
            setLoading(true);
            setError("");

            ApiController.post("subscribe", { email: email.value })
            .then(res => {
                if(res.errors) {
                    setError(res.errors[0].message);
                } else {
                    setSuccess(res.data.message);
                }

                setLoading(false);
            });
        }
    };

    if(loading) {
        return (
            <div className="main-footer__subscription-form-loader">
                <span className="spinner-border"></span>
            </div>
        );
    }

    if(!success) {
        const inputClass = error || email.error ? "formulary__input--error" : "";
        const labelClass = error || email.error ? "formulary__error-label--error" : "";

        return (
            <form onSubmit={suscribe}>
                <div className="formulary__input-container">
                    <input
                    id={`subscribe-input`}
                    type="email"
                    className={`formulary__input ${inputClass}`}
                    onChange={({ target: { value } }) => setEmail(value)}
                    value={email.value}
                    placeholder="Enter your email"
                    autoComplete="email"
                    maxLength="100"
                    required/>

                    <label
                    className={`formulary__error-label ${labelClass} mt-2`}
                    htmlFor={`subscribe-input`}>
                        <i className="fas fa-info-circle mr-1" aria-hidden="true"></i>
                        { error || email.error }
                    </label>
                </div>

                <button
                className="formulary__submit-button mt-3"
                disabled={loading}>
                    Suscribe
                </button>
            </form>
        );
    }
    
    return (
        <p className="main-footer__text m-0">
            { success }
        </p>
    );
}