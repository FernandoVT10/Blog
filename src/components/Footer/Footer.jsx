import Link from "next/link";

import SubscribeForm from "./SubscribeForm";

import "./Footer.scss";

export default () => {
    return (
        <footer className="main-footer container-fluid mt-4">

            <div className="row">
                <div className="col-12 col-sm-6 col-xl-3">
                    <Link href="/">
                        <a className="main-footer__brand">
                            <img
                            src="/favicon.ico"
                            width="50"
                            height="50"
                            className="d-inline-block align-top rounded-circle"
                            alt="Icon" />
                            <h3 className="main-footer__title">
                                Fernando Vaca Tamayo
                            </h3>
                        </a>
                    </Link>

                    <div className="mt-4">
                        <h3 className="main-footer__subtitle">
                            Contact Me
                        </h3>

                        <p>
                            <i
                            className="fas fa-envelope main-footer__icon"
                            aria-hidden="true"></i>

                            <span className="main-footer__text">
                                vacatamayo321@gmail.com
                            </span>
                        </p>
                    </div>
                </div>

                <div className="col-12 col-sm-6 col-xl-3">
                    <h3 className="main-footer__subtitle">
                        Social Networks
                    </h3>

                    <a href="https://twitter.com/FernandoVT10" target="_blank">
                        <p>
                            <i
                            className="fab fa-twitter main-footer__icon"
                            aria-hidden="true"></i>

                            <span className="main-footer__text">
                                @FernandoVT10
                            </span>
                        </p>
                    </a>

                    <a href="https://github.com/FernandoVT10" target="_blank">
                        <p>
                            <i
                            className="fab fa-github main-footer__icon"
                            aria-hidden="true"></i>

                            <span className="main-footer__text">
                                FernandoVT10
                            </span>
                        </p>
                    </a>
                </div>

                <div className="col-12 col-sm-6 col-xl-3">
                    <h3 className="main-footer__subtitle">
                        Description
                    </h3>

                    <p className="main-footer__description">
                        Hello, i am Fernando Vaca Tamayo.<br/>
                        In this blog you'll find technology tutorials.
                    </p>
                </div>

                <div className="col-12 col-sm-6 col-xl-3">
                    <h3 className="main-footer__subtitle">
                        Suscribe to my blog
                    </h3>

                    <SubscribeForm/>
                </div>
            </div>
        </footer>
    );
};