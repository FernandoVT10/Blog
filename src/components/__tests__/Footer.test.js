import Footer from "../Footer/";
import { act, Simulate } from "react-dom/test-utils";
import { render } from "react-dom";
import { useRouter } from "next/router";
import { useState } from "react";

window.$ = () => ({ on: jest.fn(), modal: jest.fn() })

jest.mock("next/router", () => ({
    useRouter: jest.fn(() => ({ query: {} })),
}));

let container;

function T() {
    const [name, setName] = useState("");

    const onClick = () => {
        fetch("https://asdasdas.com")
        .then(res => res.json())
        .then(data => setName("Hola"));
    };

    return <button onClick={onClick}><span>{name}</span></button>;
}

describe('<ArticleFilter/> Component', () => {
    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
    });
    
    afterEach(() => {
        document.body.removeChild(container);
        container = null;
    });

    it('Check if renders correctly', () => {
        act(() => {
            render(<Footer/>, container);
        });

        expect(container).toMatchSnapshot();
    });

    it('It should display a success message on the suscribe form', async () => {
        fetchMock.doMock();

        fetchMock.mockOnce(JSON.stringify(
            { status: true, message: "Test success message" }
        ));

        render(<Footer/>, container);

        const form = container.querySelector("form");
        const input = container.querySelector("input");

        act(() => {
            input.value = "test@gmail.com";
            Simulate.change(input);
        });

        await act(async () => {
            Simulate.submit(form);
        });
        
        const p = container.querySelector(".main-footer__text.m-0");

        expect(p.textContent).toBe("Test success message");
    });

    it("It should show a success message on a modal", async () => {
        fetchMock.doMock();

        fetchMock.mockOnce(JSON.stringify(
            { status: true, message: "Test success message" }
        ));

        useRouter.mockImplementation(() => ({
            query: { subscriptionId: "a12af43f2a2b345c2a2fe" }
        }));

        await act(async () => {
            render(<Footer/>, container); 
        });

        const p = container.querySelector(".modal__body p");

        expect(p.textContent).toBe("Test success message");
    });
});