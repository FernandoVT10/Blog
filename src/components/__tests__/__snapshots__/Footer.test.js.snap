// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleFilter/> Component Check if renders correctly 1`] = `
<div>
  <footer
    class="main-footer container-fluid mt-4"
  >
    <div
      class="full-screen-loader "
    >
      <div
        class="full-screen-loader__spinner spinner-border text-light"
        role="status"
      >
        <span
          class="sr-only"
        >
          Loading...
        </span>
      </div>
      <h3
        class="full-screen-loader__text"
      >
        Loading...
      </h3>
    </div>
    <div
      class="modal fade"
      id="-modal"
      role="dialog"
      tabindex="-1"
    >
      <div
        class="modal-dialog "
        role="document"
      >
        <div
          class="modal-content modal__content"
        >
          <div
            class="modal-header modal__header"
          >
            <h5
              class="modal__title"
            >
              Suscription confirm status
            </h5>
            <button
              aria-label="Close"
              class="close modal__button"
              data-dismiss="modal"
              type="button"
            >
              <span
                aria-hidden="true"
              >
                Ã—
              </span>
            </button>
          </div>
          <div
            class="modal-body modal__body"
          >
            <p />
          </div>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-12 col-sm-6 col-xl-3"
      >
        <a
          class="main-footer__brand"
          href="/"
        >
          <img
            alt="Icon"
            class="d-inline-block align-top rounded-circle"
            height="50"
            src="/favicon.ico"
            width="50"
          />
          <h3
            class="main-footer__title"
          >
            Fernando Vaca Tamayo
          </h3>
        </a>
        <div
          class="mt-4"
        >
          <h3
            class="main-footer__subtitle"
          >
            Contact Me
          </h3>
          <p>
            <i
              aria-hidden="true"
              class="fas fa-envelope main-footer__icon"
            />
            <span
              class="main-footer__text"
            >
              vacatamayo321@gmail.com
            </span>
          </p>
        </div>
      </div>
      <div
        class="col-12 col-sm-6 col-xl-3"
      >
        <h3
          class="main-footer__subtitle"
        >
          Social Networks
        </h3>
        <a
          href="https://twitter.com/FernandoVT10"
          target="_blank"
        >
          <p>
            <i
              aria-hidden="true"
              class="fab fa-twitter main-footer__icon"
            />
            <span
              class="main-footer__text"
            >
              @FernandoVT10
            </span>
          </p>
        </a>
        <a
          href="https://github.com/FernandoVT10"
          target="_blank"
        >
          <p>
            <i
              aria-hidden="true"
              class="fab fa-github main-footer__icon"
            />
            <span
              class="main-footer__text"
            >
              FernandoVT10
            </span>
          </p>
        </a>
      </div>
      <div
        class="col-12 col-sm-6 col-xl-3"
      >
        <h3
          class="main-footer__subtitle"
        >
          Description
        </h3>
        <p
          class="main-footer__description"
        >
          This blog is my personal blog xD
        </p>
      </div>
      <div
        class="col-12 col-sm-6 col-xl-3"
      >
        <h3
          class="main-footer__subtitle"
        >
          Suscribe to my blog
        </h3>
        <form>
          <div
            class="formulary__input-container"
          >
            <input
              autocomplete="email"
              class="formulary__input "
              id="-input"
              maxlength="100"
              placeholder="Enter your email"
              required=""
              type="email"
            />
            <label
              class="formulary__error-label mt-2 "
              for="-input"
            >
              <i
                aria-hidden="true"
                class="fas fa-info-circle mr-1"
              />
              
            </label>
          </div>
          <button
            class="submit-button mt-3"
          >
            Suscribe
          </button>
        </form>
      </div>
    </div>
  </footer>
</div>
`;
